src/pywifes/extract_spec.py

```diff
@@ -0,0 +1,516 @@
+        Aperture object representing the source region to be extracted.
+
+    sky_ap : `photutils.aperture` object, optional
+        Aperture object representing the sky region to be extracted.
+        Default is None.
+
+    Returns
+    -------
+    flux : ndarray
+        1D array containing the spectral flux values.
+
+    variance : ndarray
+        1D array containing the spectral variance values.
+
+    Notes
+    -----
+    This function extracts spectral flux and variance values within the specified
+    source aperture from the input data cubes. If a sky aperture is provided,
+    it calculates and subtracts the corresponding sky values.
+
+    """
+    # Extract the pixels inside the aperture for all the wavelentghs
+    fl = []
+    var = []
+
+    for layer in range(sci_cube.shape[0]):
+        sci_section = sec_image(source_ap, sci_cube[layer])
+        var_section = sec_image(source_ap, var_cube[layer])
+        area = source_ap.area_overlap(sci_cube[layer], method="center")
+
+        average = np.average(
+            sci_section,
+            weights=np.reciprocal(var_section),
+        )
+
+        error = np.reciprocal(np.sum(np.reciprocal(var_section)))
+
+        if sky_ap is not None:
+            sky_section = sec_image(sky_ap, sci_cube[layer])
+            sky_var_section = sec_image(sky_ap, var_cube[layer])
+
+            sky_average = np.average(
+                sky_section,
+                weights=np.reciprocal(sky_var_section),
+            )
+
+            sky_error = np.reciprocal(np.sum(np.reciprocal(sky_var_section)))
+
+            fl.append((average - sky_average) * area)
+            var.append((error + sky_error) * area)
+
+        else:
+            fl.append(average * area)
+            var.append(error * area)
+
+    return np.array(fl), np.array(var)
+
+
+def write_1D_spec(sci_data, var_data, sci_cube_header, var_cube_header, output):
+    """
+    This function writes the 1D spectrum data and its variance to a FITS file.
+    It updates the WCS headers to match the 1D spectrum header.
+
+    Parameters
+    ----------
+    sci_data : numpy.ndarray
+        1D array containing the science data (flux) of the spectrum.
+    var_data : numpy.ndarray
+        1D array containing the variance data of the spectrum.
+    sci_cube_header : astropy.io.fits.Header
+        Header from the science cube.
+    var_cube_header : astropy.io.fits.Header
+        Header object from the variance data.
+    output : str
+        Output file path for the FITS file containing the 1D spectrum.
+
+    Returns
+    -------
+    None
+        This function does not return any value. It writes the corresponding fits.
+
+    """
+
+    # Create a blank HDUList for the 1D spectrum
+    hdulist = fits.HDUList(fits.PrimaryHDU())
+
+    # Update WCS headers to match the 1D spectrum header
+    sci_cube_header_copy = sci_cube_header.copy()
+    var_cube_header_copy = var_cube_header.copy()
```

Comment:
Ahhh right, yes makes sense that you'd need a copy, and it makes sense intuitively that you wouldn't expect the header you pass in to be overwritten.